src/extra/components/dld.py

```diff
@@ -0,0 +1,304 @@
+                         if np.issubdtype(dtype, np.floating)}.pop()
+
+                finite_mask = np.isfinite(raw[field])
+            elif np.issubdtype(data.dtype, np.floating):
+                finite_mask = np.isfinite(raw)
+            else:
+                raise TypeError(data.dtype)
+
+            # Obtain pulse index and repeat each row by the number of
+            # actual data in each entry, then convert to a dataframe.
+            index_df = pulse_index.repeat(
+                finite_mask.reshape(-1, num_rows).sum(axis=1)).to_frame()
+
+            # Insert additional index level enumerating actual entries.
+            if entry_level is not None:
+                index_df[entry_level] = np.flatnonzero(finite_mask) % num_rows
+
+            res.append(pd_cls(
+                raw[finite_mask], pd.MultiIndex.from_frame(index_df)))
+
+        return res if split_axis is not None else res[0]
+
+    @property
+    def detector_name(self):
+        return self._detector_name
+
+    @property
+    def control_source(self):
+        """Control source."""
+
+        if self._control_src is None:
+            raise ValueError('component is initialized with earlier data '
+                             'lacking a control source')
+
+        return self._control_src
+
+    @property
+    def instrument_source(self):
+        """Instrument source."""
+        return self._instrument_src
+
+    @property
+    def rec_params(self):
+        """Reconstruction parameters."""
+        return self.control_source.run_values()
+
+    def select_trains(self, trains):
+        new_self = copy(self)
+
+        if self._control_src is not None:
+            new_self._control_src = self._control_src.select_trains(trains)
+        new_self._instrument_src = self._instrument_src.select_trains(trains)
+        new_self._pulses = self._pulses.select_trains(trains)
+
+        return new_self
+
+    def get_pulses(self):
+        """Get pulse object based on internal triggers.
+
+        Returns:
+            (extra.components.DldPulses): Pulse object based on
+                constructed trigger information.
+        """
+
+        from .pulses import DldPulses
+
+        if isinstance(self._pulses, DldPulses):
+            return self._pulses
+
+        return DldPulses(self._instrument_src)
+
+    def get_triggers(self):
+        """Get triggers as dataframe.
+
+        Returns:
+            (pandas.DataFrame): Constructed trigger information as
+                dataframe.
+        """
+
+        return self.get_pulses().get_triggers()
+
+    def get_edges(self, channel_index=True, pulse_dim='pulseId'):
+        """Get raw edges as series or dataframe.
+
+        Args:
+            channel_index (bool, optional): Whether to insert the edge
+                channel as index level and return Series object
+                (default), or as column and return DataFrame object.
+            pulse_dim ({pulseId, pulseIndex, time}, optional): Label
```

Comment:
I'm hazy on the pulse naming plans, but is the idea that eventually `pulseId` will be something different from `pulseIndex`? Because if they mean the same thing I think we should stick to a single name.

Reply:
Yes, `pulseId` is the index in the bunch pattern table and something we try to establish as a *global* pulse ID across the facility. `pulseIndex` is an enumeration for any beamline's pulses.

Reply:
Hmm, kinda annoying since there's a lot of code already using `pulseId` but ok :frowning_face: But anyway, do I understand correctly that this is capable of retrieving the edges labelled by either `pulseId` or `pulseIndex` which will be different numbers but refer to the same pulse? If so, what does `time` mean? Is that something coming from the DLD?

Reply:
Unfortunately yes, all kind of equipment has historically just made up its own pulse IDs. The big detectors are particularly big offenders, as they're not even enumerating pulses, but their own frames. Some experiments take frames between pulses, making the mess complete.

When writing the pulse components, I found it hard to use _id_ for something that's arbitrarily enumerated, and thus chose _index_.

These arguments are directly handed off to `PulsePattern.build_pulse_index`, and yes, the aim is to allow different ways how to label pulses. `time` is just that, the actual time of the pulse relative to the beginning of the train. The idea behind this is that some experiments (e.g. at HED) use pulse trains as a means of probing the same at different _times_.

I will extend the documentation a bit more to illuminate this, thanks! 🙂